src/view/treewalker.js

```diff
@@ -0,0 +1,392 @@
+		 *
+		 * @readonly
+		 * @member {Boolean} engine.view.TreeWalker#singleCharacters
+		 */
+		this.singleCharacters = !!options.singleCharacters;
+
+		/**
+		 * Flag indicating whether iterator should enter elements or not. If the iterator is shallow child nodes of any
+		 * iterated node will not be returned along with `ELEMENT_END` tag.
+		 *
+		 * @readonly
+		 * @member {Boolean} engine.view.TreeWalker#shallow
+		 */
+		this.shallow = !!options.shallow;
+
+		/**
+		 * Flag indicating whether iterator should ignore `ELEMENT_END` tags. If the option is true walker will not
+		 * return a parent node of the start position. If this option is `true` each {@link engine.view.Element} will
+		 * be returned once, while if the option is `false` they might be returned twice:
+		 * for `'ELEMENT_START'` and `'ELEMENT_END'`.
+		 *
+		 * @readonly
+		 * @member {Boolean} engine.view.TreeWalker#ignoreElementEnd
+		 */
+		this.ignoreElementEnd = !!options.ignoreElementEnd;
+
+		/**
+		 * Start boundary cached for optimization purposes.
+		 *
+		 * @private
+		 * @member {engine.view.Element} engine.view.TreeWalker#_boundaryStartParent
+		 */
+		this._boundaryStartParent = this.boundaries ? this.boundaries.start.parent : null;
+
+		/**
+		 * End boundary cached for optimization purposes.
+		 *
+		 * @private
+		 * @member {engine.view.Element} engine.view.TreeWalker#_boundaryEndParent
+		 */
+		this._boundaryEndParent = this.boundaries ? this.boundaries.end.parent : null;
+	}
+
+	/**
+	 * Iterator interface.
+	 */
+	[ Symbol.iterator ]() {
+		return this;
+	}
+
+	/**
+	 * Iterator interface method.
+	 * Detects walking direction and makes step forward or backward.
+	 *
+	 * @returns {Object} Object implementing iterator interface, returning information about taken step.
+	 */
+	next() {
+		if ( this.direction == 'FORWARD' ) {
+			return this._next();
+		} else {
+			return this._previous();
+		}
+	}
+
+	/**
+	 * Makes a step forward in view. Moves the {@link #position} to the next position and returns the encountered value.
+	 *
+	 * @private
+	 * @returns {Object}
+	 * @returns {Boolean} return.done True if iterator is done.
+	 * @returns {engine.view.TreeWalkerValue} return.value Information about taken step.
+	 */
+	_next() {
+		let position = Position.createFromPosition( this.position );
+		const previousPosition = this.position;
+		const parent = position.parent;
+
+		// We are at the end of the root.
+		if ( parent.parent === null && position.offset === parent.getChildCount() ) {
+			return { done: true };
+		}
+
+		// We reached the walker boundary.
+		if ( parent === this._boundaryEndParent && position.offset == this.boundaries.end.offset ) {
+			return { done: true };
+		}
+
+		// Get node just after current position.
+		let node;
+
+		// Text {@link engine.view.Text} element is a specific parent because it contains string instead of child nodes.
+		if ( parent instanceof Text ) {
+			node = parent._data[ position.offset ];
```

Comment:
There is a bug. If you start in the middle of the text node, but without the singleCharacters option, the node is a letter anyway. And then, when the node is string, you return single characters till the end of the string. Test does not couch it, because in the test, the position starts one letter before the end of the text node, so it is fine there.